component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -0,0 +1,202 @@
+        }
+
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticating client : " + oAuthClientAuthnContext.getClientId() + " with public " +
+                        "certificate.");
+            }
+
+            X509Certificate requestCert;
+            Object certObject = request.getAttribute(JAVAX_SERVLET_REQUEST_CERTIFICATE);
+            if (certObject instanceof X509Certificate[]) {
+                X509Certificate[] cert = (X509Certificate[]) certObject;
+                requestCert = cert[0];
+            } else if (certObject instanceof X509Certificate){
+                requestCert = (X509Certificate) certObject;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Could not find client certificate in required format.");
+                }
+                return false;
+            }
+
+            String tenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthClientAuthnContext.getClientId());
+            X509Certificate registeredCert = (X509Certificate) OAuth2Util
+                    .getX509CertOfOAuthApp(oAuthClientAuthnContext.getClientId(), tenantDomain);
+            return authenticate(registeredCert, requestCert);
+
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR, "Error occurred while retrieving " +
+                    "public certificate of client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (InvalidOAuthClientException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.INVALID_CLIENT, "Error occurred while retrieving " +
+                    "tenant domain for the client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        }
+    }
+
+    /**
+     * Returns whether the incoming request can be authenticated or not using the given inputs.
+     *
+     * @param request    HttpServletRequest which is the incoming request.
+     * @param bodyParams Body parameters present in the request.
+     * @param context    OAuth2 client authentication context.
+     * @return Whether client can be authenticated using this authenticator.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
+                                   OAuthClientAuthnContext context) {
+
+        if (clientIdExistsAsParam(bodyParams) && validCertExistsAsAttribute(request)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Client ID exists in request body parameters and a valid certificate found in request " +
+                        "attributes. Hence returning true.");
+            }
+            return true;
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Client id is not available in body params or valid certificate not found in request " +
+                        "attributes. Hence returning false.");
```

Comment:
Rather than saying "returning false", may be you can write "Mutual TLS authenticator cannot handle this request". Because a person who reads this is not aware of code. So he should not be able to identify for what actually this is returning false.